File: camera/capture.go
package main

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"
	"time"
)

const segmentDurationSecs = 6

// StartCapturePipeline spawns the capture pipeline and blocks until it exits or
// the context is cancelled. For test-source mode it uses ffmpeg testsrc2; for
// real hardware it pipes rpicam-vid into ffmpeg.
//
// The liveWriter receives a copy of the raw H.264 bytestream for WebRTC live
// relay. Audio Opus packets are pushed via liveWriter.PushAudio from the OGG
// reader goroutine. Pass NullLiveRelay{} to disable live streaming.
//
// When cfg.RecordingMode == "never" the pipeline still runs — the live relay
// output (raw H.264 on stdout + OGG/Opus on fd 3) keeps working — but the
// MPEG-TS segment sink is omitted so nothing is ever written to SegmentDir.
func StartCapturePipeline(ctx context.Context, cfg *CameraConfig, liveWriter LiveWriter) error {
	startNum := nextSegmentNumber(cfg.SegmentDir)
	pattern := filepath.Join(cfg.SegmentDir, "seg%05d.ts")
	kfInterval := fmt.Sprintf("keyint=%d:min-keyint=%d", cfg.VideoKeyframeInterval, cfg.VideoKeyframeInterval)

	if cfg.TestSource {
		return runTestPipeline(ctx, cfg, pattern, kfInterval, startNum, liveWriter)
	}
	return runRealPipeline(ctx, cfg, pattern, startNum, liveWriter)
}

// recordsSegments reports whether the pipeline should emit MPEG-TS segments
// for HLS upload. Streaming-only mode suppresses the segment sink.
func recordsSegments(cfg *CameraConfig) bool {
	return cfg.RecordingMode != "never"
}

// nextSegmentNumber counts existing .ts files to avoid filename collisions on restart.
func nextSegmentNumber(dir string) int {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return 0
	}
	count := 0
	for _, e := range entries {
		if filepath.Ext(e.Name()) == ".ts" {
			count++
		}
	}
	return count
}

func runTestPipeline(ctx context.Context, cfg *CameraConfig, pattern, kfInterval string, startNum int, liveWriter LiveWriter) error {
	// Prefer pre-encoded test file (no CPU-intensive encoding)
	testFile := filepath.Join(cfg.DataDir, "test-loop.mp4")
	if _, err := os.Stat(testFile); err == nil {
		return runTestFileLoop(ctx, cfg, testFile, pattern, startNum, liveWriter)
	}

	slog.Info("starting test capture pipeline (ffmpeg testsrc2 + sine audio)", "segment_start", startNum)

	size := fmt.Sprintf("%dx%d", cfg.VideoWidth, cfg.VideoHeight)
	videoInput := fmt.Sprintf("testsrc2=size=%s:rate=%d,drawtext=fontfile=/usr/share/fonts/dejavu/DejaVuSansMono.ttf:text='%%{localtime\\:%%T}':fontsize=48:fontcolor=white:x=10:y=10", size, cfg.VideoFPS)
	audioInput := "sine=frequency=440:sample_rate=48000"

	// Create pipe for Opus audio output (fd 3 inside ffmpeg).
	audioPipeR, audioPipeW, err := os.Pipe()
	if err != nil {
		return fmt.Errorf("creating audio pipe: %w", err)
	}
	defer audioPipeR.Close()

	// ffmpeg outputs:
	//   0: MPEG-TS segments (H.264 + AAC) — for HLS recording (omitted in "never" mode)
	//   1: raw H.264 to stdout — for WebRTC video
	//   2: OGG/Opus to fd 3 — for WebRTC audio
	args := []string{
		"-re",
		"-f", "lavfi", "-i", videoInput,
		"-f", "lavfi", "-i", audioInput,
	}
	if recordsSegments(cfg) {
		args = append(args,
			// Output 0: MPEG-TS segments (video + audio)
			"-map", "0:v", "-map", "1:a",
			"-c:v", "libx264",
			"-preset", "ultrafast",
			"-x264-params", kfInterval,
			"-c:a", "aac", "-b:a", "64k",
			"-f", "segment",
			"-segment_time", fmt.Sprintf("%d", segmentDurationSecs),
			"-segment_format", "mpegts",
			"-segment_start_number", fmt.Sprintf("%d", startNum),
			"-reset_timestamps", "1",
			pattern,
		)
	}
	args = append(args,
		// Output 1: raw H.264 to stdout for WebRTC video
		"-map", "0:v",
		"-c:v", "libx264",
		"-preset", "ultrafast",
		"-x264-params", kfInterval,
		"-f", "h264",
		"pipe:1",
		// Output 2: OGG/Opus to fd 3 for WebRTC audio
		"-map", "1:a",
		"-c:a", "libopus", "-b:a", "32k",
		"-application", "lowdelay",
		"-frame_duration", "20",
		"-f", "ogg",
		"pipe:3",
	)
	cmd := exec.CommandContext(ctx, "ffmpeg", args...)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		audioPipeW.Close()
		return fmt.Errorf("creating stdout pipe: %w", err)
	}

	// Pass the audio pipe write end as fd 3.
	cmd.ExtraFiles = []*os.File{audioPipeW}

	if err := cmd.Start(); err != nil {
		audioPipeW.Close()
		return fmt.Errorf("starting ffmpeg: %w", err)
	}
	// Close the write end in the parent — ffmpeg owns it now.
	audioPipeW.Close()
	slog.Info("ffmpeg test pipeline started", "pid", cmd.Process.Pid)

	// Copy stdout (raw H.264) to the live writer.
	go func() {
		io.Copy(liveWriter, stdout)
	}()

	// Read OGG/Opus from the audio pipe and push to the live relay.
	go func() {
		if err := ReadOggOpusPackets(audioPipeR, liveWriter.PushAudio); err != nil {
			slog.Debug("opus reader finished", "err", err)
		}
	}()

	err = cmd.Wait()
	if ctx.Err() != nil {
		slog.Info("capture pipeline cancelled")
		return ctx.Err()
	}
	if err != nil {
		return fmt.Errorf("ffmpeg exited: %w", err)
	}
	return nil
}

// runTestFileLoop loops a pre-encoded MP4 file with -c copy (no encoding, minimal CPU).
func runTestFileLoop(ctx context.Context, cfg *CameraConfig, testFile, pattern string, startNum int, liveWriter LiveWriter) error {
	slog.Info("starting test capture pipeline (pre-encoded loop, -c copy)", "file", testFile, "segment_start", startNum)

	// Create pipe for Opus audio output.
	audioPipeR, audioPipeW, err := os.Pipe()
	if err != nil {
		return fmt.Errorf("creating audio pipe: %w", err)
	}
	defer audioPipeR.Close()

	args := []string{
		"-re",
		"-stream_loop", "-1",
		"-i", testFile,
	}
	if recordsSegments(cfg) {
		args = append(args,
			// Output 0: segments
			"-map", "0",
			"-c", "copy",
			"-f", "segment",
			"-segment_time", fmt.Sprintf("%d", segmentDurationSecs),
			"-segment_format", "mpegts",
			"-segment_start_number", fmt.Sprintf("%d", startNum),
			pattern,
		)
	}
	args = append(args,
		// Output 1: raw H.264 to stdout
		"-map", "0:v",
		"-c:v", "copy",
		"-f", "h264",
		"pipe:1",
		// Output 2: OGG/Opus to fd 3
		"-map", "0:a",
		"-c:a", "libopus", "-b:a", "32k",
		"-application", "lowdelay",
		"-frame_duration", "20",
		"-f", "ogg",
		"pipe:3",
	)
	cmd := exec.CommandContext(ctx, "ffmpeg", args...)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		audioPipeW.Close()
		return fmt.Errorf("creating stdout pipe: %w", err)
	}

	cmd.ExtraFiles = []*os.File{audioPipeW}

	if err := cmd.Start(); err != nil {
		audioPipeW.Close()
		return fmt.Errorf("starting ffmpeg: %w", err)
	}
	audioPipeW.Close()
	slog.Info("ffmpeg test file loop started", "pid", cmd.Process.Pid)

	go func() {
		io.Copy(liveWriter, stdout)
	}()

	go func() {
		if err := ReadOggOpusPackets(audioPipeR, liveWriter.PushAudio); err != nil {
			slog.Debug("opus reader finished", "err", err)
		}
	}()

	err = cmd.Wait()
	if ctx.Err() != nil {
		slog.Info("capture pipeline cancelled")
		return ctx.Err()
	}
	if err != nil {
		return fmt.Errorf("ffmpeg exited: %w", err)
	}
	return nil
}

func runRealPipeline(ctx context.Context, cfg *CameraConfig, pattern string, startNum int, liveWriter LiveWriter) error {
	hasAudio := !cfg.NoAudio
	record := recordsSegments(cfg)
	slog.Info("starting real capture pipeline",
		"segment_start", startNum, "records_segments", record, "has_audio", hasAudio)

	// --- rpicam-vid: produces raw H.264 Annex B on stdout ---
	rpicamCmd := exec.CommandContext(ctx, "rpicam-vid",
		"--codec", "h264",
		"--inline",
		"--width", fmt.Sprintf("%d", cfg.VideoWidth),
		"--height", fmt.Sprintf("%d", cfg.VideoHeight),
		"--framerate", fmt.Sprintf("%d", cfg.VideoFPS),
		"--bitrate", fmt.Sprintf("%d", cfg.VideoBitrate),
		"-t", "0",
		"-o", "-",
	)
	rpicamCmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	rpicamStdout, err := rpicamCmd.StdoutPipe()
	if err != nil {
		return fmt.Errorf("rpicam stdout pipe: %w", err)
	}

	// Streaming-only with no audio: ffmpeg has nothing to do. Pipe
	// rpicam-vid straight into the live writer and return.
	if !record && !hasAudio {
		return runLiveOnlyRealPipeline(ctx, rpicamCmd, rpicamStdout, liveWriter)
	}

	// feedVideoToFFmpeg controls whether H.264 is piped into ffmpeg. We only
	// do so when segments are being recorded. In streaming-only mode with
	// audio we still need ffmpeg for the ALSA→Opus path, but piping H.264
	// into an unmapped input would back-pressure the rpicam-vid tee and
	// stall the live relay — so rpicam-vid's stdout goes straight to
	// liveWriter and ffmpeg only sees the ALSA input.
	feedVideoToFFmpeg := record

	// Stream index of the ALSA input inside ffmpeg. When H.264 is also
	// piped in, ALSA is input 1; otherwise it's the only input (0).
	alsaStreamIdx := 0
	if feedVideoToFFmpeg {
		alsaStreamIdx = 1
	}

	// --- ffmpeg: muxes MPEG-TS segments (when recording) + Opus audio ---
	ffmpegArgs := []string{
		"-nostdin", "-loglevel", "warning",
		"-probesize", "5M", "-analyzeduration", "5M",
	}

	if feedVideoToFFmpeg {
		ffmpegArgs = append(ffmpegArgs,
			"-f", "h264", "-framerate", fmt.Sprintf("%d", cfg.VideoFPS),
			"-i", "pipe:0",
		)
	}

	var audioPipeR, audioPipeW *os.File

	if hasAudio {
		audioDevice := "default"
		if cfg.AudioDevice != "" {
			audioDevice = cfg.AudioDevice
		}
		ffmpegArgs = append(ffmpegArgs, "-f", "alsa", "-i", audioDevice)
	}

	if record {
		if hasAudio {
			ffmpegArgs = append(ffmpegArgs,
				"-map", "0:v", "-map", "1:a",
				"-c:v", "copy", "-c:a", "aac", "-b:a", "64k",
			)
		} else {
			ffmpegArgs = append(ffmpegArgs,
				"-map", "0:v",
				"-c:v", "copy",
			)
		}
		ffmpegArgs = append(ffmpegArgs,
			"-f", "segment",
			"-segment_time", fmt.Sprintf("%d", segmentDurationSecs),
			"-segment_format", "mpegts",
			"-segment_start_number", fmt.Sprintf("%d", startNum),
			"-reset_timestamps", "1",
			pattern,
		)
	}

	// Opus output to fd 3 for WebRTC audio.
	if hasAudio {
		audioPipeR, audioPipeW, err = os.Pipe()
		if err != nil {
			return fmt.Errorf("creating audio pipe: %w", err)
		}
		defer audioPipeR.Close()

		ffmpegArgs = append(ffmpegArgs,
			"-map", fmt.Sprintf("%d:a", alsaStreamIdx),
			"-c:a", "libopus", "-b:a", "32k",
			"-application", "lowdelay",
			"-frame_duration", "20",
			"-f", "ogg",
			"pipe:3",
		)
	}

	ffmpegCmd := exec.CommandContext(ctx, "ffmpeg", ffmpegArgs...)
	ffmpegCmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}

	if audioPipeW != nil {
		ffmpegCmd.ExtraFiles = []*os.File{audioPipeW}
	}

	var ffmpegStdin io.WriteCloser
	if feedVideoToFFmpeg {
		ffmpegStdin, err = ffmpegCmd.StdinPipe()
		if err != nil {
			if audioPipeW != nil {
				audioPipeW.Close()
			}
			return fmt.Errorf("ffmpeg stdin pipe: %w", err)
		}
	}

	// Cancel handlers: SIGTERM the process group, then SIGKILL after 5s.
	for _, cmd := range []*exec.Cmd{rpicamCmd, ffmpegCmd} {
		c := cmd
		c.Cancel = func() error {
			pgid := -c.Process.Pid
			if err := syscall.Kill(pgid, syscall.SIGTERM); err != nil {
				return err
			}
			go func() {
				time.Sleep(5 * time.Second)
				_ = syscall.Kill(pgid, syscall.SIGKILL)
			}()
			return nil
		}
	}

	// Start both processes.
	if err := rpicamCmd.Start(); err != nil {
		return fmt.Errorf("starting rpicam-vid: %w", err)
	}
	slog.Info("rpicam-vid started", "pid", rpicamCmd.Process.Pid)

	if err := ffmpegCmd.Start(); err != nil {
		rpicamCmd.Process.Kill()
		if audioPipeW != nil {
			audioPipeW.Close()
		}
		return fmt.Errorf("starting ffmpeg: %w", err)
	}
	// Close the write end in the parent — ffmpeg owns it now.
	if audioPipeW != nil {
		audioPipeW.Close()
	}
	slog.Info("ffmpeg started", "pid", ffmpegCmd.Process.Pid)

	// Tee rpicam-vid's stdout to ffmpeg's stdin (when recording) and to
	// the live writer. In streaming-only mode ffmpegStdin is nil so the
	// live writer is the only sink for the H.264 stream.
	teeDone := make(chan error, 1)
	go func() {
		var target io.Writer = liveWriter
		if ffmpegStdin != nil {
			target = io.MultiWriter(ffmpegStdin, liveWriter)
		}
		_, err := io.Copy(target, rpicamStdout)
		if ffmpegStdin != nil {
			ffmpegStdin.Close()
		}
		teeDone <- err
	}()

	// Read Opus audio from the pipe if enabled.
	if hasAudio && audioPipeR != nil {
		go func() {
			if err := ReadOggOpusPackets(audioPipeR, liveWriter.PushAudio); err != nil {
				slog.Debug("opus reader finished", "err", err)
			}
		}()
	}

	rpicamErr := rpicamCmd.Wait()
	ffmpegErr := ffmpegCmd.Wait()
	<-teeDone

	if ctx.Err() != nil {
		slog.Info("capture pipeline cancelled")
		return ctx.Err()
	}
	if rpicamErr != nil {
		return fmt.Errorf("rpicam-vid exited: %w", rpicamErr)
	}
	if ffmpegErr != nil {
		return fmt.Errorf("ffmpeg exited: %w", ffmpegErr)
	}
	return nil
}

// runLiveOnlyRealPipeline runs rpicam-vid alone and tees its raw H.264 to the
// live writer. Used when RecordingMode == "never" and audio is disabled, since
// ffmpeg would otherwise have no outputs to produce.
func runLiveOnlyRealPipeline(ctx context.Context, rpicamCmd *exec.Cmd, rpicamStdout io.ReadCloser, liveWriter LiveWriter) error {
	rpicamCmd.Cancel = func() error {
		pgid := -rpicamCmd.Process.Pid
		if err := syscall.Kill(pgid, syscall.SIGTERM); err != nil {
			return err
		}
		go func() {
			time.Sleep(5 * time.Second)
			_ = syscall.Kill(pgid, syscall.SIGKILL)
		}()
		return nil
	}

	if err := rpicamCmd.Start(); err != nil {
		return fmt.Errorf("starting rpicam-vid: %w", err)
	}
	slog.Info("rpicam-vid started (live-only, no recording, no audio)", "pid", rpicamCmd.Process.Pid)

	teeDone := make(chan error, 1)
	go func() {
		_, err := io.Copy(liveWriter, rpicamStdout)
		teeDone <- err
	}()

	rpicamErr := rpicamCmd.Wait()
	<-teeDone

	if ctx.Err() != nil {
		slog.Info("capture pipeline cancelled")
		return ctx.Err()
	}
	if rpicamErr != nil {
		return fmt.Errorf("rpicam-vid exited: %w", rpicamErr)
	}
	return nil
}

File: ghostcam-py/ghostcam/capture.py
"""Capture pipeline: rpicam-vid + ffmpeg orchestration.

Mirrors camera/capture.go. Three modes, mirroring the Go camera:

  * test_source with pre-encoded loop file (`{data_dir}/test-loop.mp4`)
  * test_source with ffmpeg testsrc2 + sine audio
  * real rpicam-vid + ffmpeg

In all modes the capture process produces:

  * MPEG-TS segments in segment_dir (when recording_mode != "never")
  * Raw H.264 to live_writer.write() — drives the WebSocket relay
  * OGG/Opus on a side-channel pipe → live_writer.push_audio()

The OGG/Opus side-channel is the Spike 2 result: instead of pinning to
fd 3 (which Python silently breaks during interpreter startup), we tell
ffmpeg `pipe:{wfd}` for whatever fd `os.pipe()` returned and wire it in
via `pass_fds`. ffmpeg's pipe protocol accepts arbitrary fd numbers.
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import os
from pathlib import Path

from ghostcam.config import CameraConfig
from ghostcam.live_relay import LiveWriter
from ghostcam.ogg_reader import read_ogg_opus_packets

logger = logging.getLogger(__name__)

SEGMENT_DURATION_SECS = 6


def _records_segments(cfg: CameraConfig) -> bool:
    return cfg.recording_mode != "never"


def _next_segment_number(directory: Path) -> int:
    try:
        return sum(1 for e in directory.iterdir() if e.suffix == ".ts")
    except OSError:
        return 0


def _segment_pattern(segment_dir: Path) -> str:
    return str(segment_dir / "seg%05d.ts")


async def start_capture_pipeline(
    cfg: CameraConfig,
    live_writer: LiveWriter,
) -> None:
    """Spawn the capture pipeline. Returns when the pipeline exits or
    the surrounding task is cancelled. Raises on subprocess failure.
    """
    cfg.segment_dir.mkdir(parents=True, exist_ok=True)
    start_num = _next_segment_number(cfg.segment_dir)
    pattern = _segment_pattern(cfg.segment_dir)

    if cfg.test_source:
        loop_file = cfg.data_dir / "test-loop.mp4"
        if loop_file.is_file():
            await _run_test_file_loop(cfg, loop_file, pattern, start_num, live_writer)
        else:
            await _run_test_pipeline(cfg, pattern, start_num, live_writer)
        return
    await _run_real_pipeline(cfg, pattern, start_num, live_writer)


# --- helpers ---


async def _spawn_ffmpeg_with_audio_pipe(
    args: list[str],
    has_audio: bool,
) -> tuple[asyncio.subprocess.Process, asyncio.StreamReader | None]:
    """Spawn ffmpeg and, when has_audio, set up the side-channel pipe.

    Returns (proc, audio_reader). audio_reader is None when has_audio is
    False. Caller must ensure args contain the matching `pipe:{wfd}`
    output spec (the wfd is allocated here and we substitute the literal
    placeholder `{audio_fd}` in args).
    """
    audio_pipe_r: asyncio.StreamReader | None = None
    pass_fds: tuple[int, ...] = ()

    if has_audio:
        rfd, wfd = os.pipe()
        os.set_inheritable(wfd, True)
        # Substitute placeholder in args.
        args = [a.replace("{audio_fd}", str(wfd)) for a in args]
        pass_fds = (wfd,)

        proc = await asyncio.create_subprocess_exec(
            *args,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            pass_fds=pass_fds,
        )
        os.close(wfd)

        loop = asyncio.get_running_loop()
        audio_pipe_r = asyncio.StreamReader()
        protocol = asyncio.StreamReaderProtocol(audio_pipe_r)
        await loop.connect_read_pipe(lambda: protocol, os.fdopen(rfd, "rb"))
        return proc, audio_pipe_r

    proc = await asyncio.create_subprocess_exec(
        *args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    return proc, None


async def _drain_stderr(proc: asyncio.subprocess.Process, label: str) -> None:
    if proc.stderr is None:
        return
    while True:
        line = await proc.stderr.readline()
        if not line:
            return
        logger.debug("%s: %s", label, line.decode(errors="replace").rstrip())


async def _copy_to_live(
    src: asyncio.StreamReader,
    live_writer: LiveWriter,
    *,
    chunk: int = 65536,
) -> None:
    while True:
        data = await src.read(chunk)
        if not data:
            return
        live_writer.write(data)


async def _copy_to_two(
    src: asyncio.StreamReader,
    sinks: list[asyncio.StreamWriter | None],
    live_writer: LiveWriter,
    *,
    chunk: int = 65536,
) -> None:
    """Copy src bytes to each non-None StreamWriter sink AND to the live
    writer. Closes the StreamWriter sinks on EOF.
    """
    try:
        while True:
            data = await src.read(chunk)
            if not data:
                return
            live_writer.write(data)
            for w in sinks:
                if w is None:
                    continue
                w.write(data)
                try:
                    await w.drain()
                except (BrokenPipeError, ConnectionResetError):
                    return
    finally:
        for w in sinks:
            if w is None:
                continue
            with contextlib.suppress(Exception):
                w.close()


async def _audio_reader_task(
    audio_pipe: asyncio.StreamReader,
    live_writer: LiveWriter,
) -> None:
    try:
        await read_ogg_opus_packets(audio_pipe, live_writer.push_audio)
    except Exception as e:  # noqa: BLE001
        logger.debug("opus reader finished: %s", e)


# --- test-source pipelines ---


async def _run_test_pipeline(
    cfg: CameraConfig,
    pattern: str,
    start_num: int,
    live_writer: LiveWriter,
) -> None:
    logger.info("starting test capture pipeline (testsrc2 + sine)")

    size = f"{cfg.video_width}x{cfg.video_height}"
    video_input = (
        f"testsrc2=size={size}:rate={cfg.video_fps},"
        "drawtext=fontfile=/usr/share/fonts/dejavu/DejaVuSansMono.ttf:"
        "text='%{localtime\\:%T}':fontsize=48:fontcolor=white:x=10:y=10"
    )
    audio_input = "sine=frequency=440:sample_rate=48000"
    kf = f"keyint={cfg.video_keyframe_interval}:min-keyint={cfg.video_keyframe_interval}"

    args = [
        "ffmpeg",
        "-re",
        "-f", "lavfi", "-i", video_input,
        "-f", "lavfi", "-i", audio_input,
    ]
    if _records_segments(cfg):
        args += [
            "-map", "0:v", "-map", "1:a",
            "-c:v", "libx264", "-preset", "ultrafast",
            "-x264-params", kf,
            "-c:a", "aac", "-b:a", "64k",
            "-f", "segment",
            "-segment_time", str(SEGMENT_DURATION_SECS),
            "-segment_format", "mpegts",
            "-segment_start_number", str(start_num),
            "-reset_timestamps", "1",
            pattern,
        ]
    args += [
        "-map", "0:v",
        "-c:v", "libx264", "-preset", "ultrafast",
        "-x264-params", kf,
        "-f", "h264", "pipe:1",
        "-map", "1:a",
        "-c:a", "libopus", "-b:a", "32k",
        "-application", "lowdelay",
        "-frame_duration", "20",
        "-f", "ogg", "pipe:{audio_fd}",
    ]

    proc, audio_pipe = await _spawn_ffmpeg_with_audio_pipe(args, has_audio=True)
    logger.info("ffmpeg test pipeline started (pid=%d)", proc.pid)

    tasks = [
        asyncio.create_task(_copy_to_live(proc.stdout, live_writer), name="ffmpeg-stdout"),  # type: ignore[arg-type]
        asyncio.create_task(_drain_stderr(proc, "ffmpeg"), name="ffmpeg-stderr"),
    ]
    if audio_pipe is not None:
        tasks.append(asyncio.create_task(
            _audio_reader_task(audio_pipe, live_writer), name="opus-reader"
        ))

    try:
        rc = await proc.wait()
    finally:
        for t in tasks:
            t.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)

    if rc != 0:
        raise RuntimeError(f"ffmpeg exited with code {rc}")


async def _run_test_file_loop(
    cfg: CameraConfig,
    test_file: Path,
    pattern: str,
    start_num: int,
    live_writer: LiveWriter,
) -> None:
    logger.info("starting test file loop (-c copy): %s", test_file)

    args = [
        "ffmpeg",
        "-re",
        "-stream_loop", "-1",
        "-i", str(test_file),
    ]
    if _records_segments(cfg):
        args += [
            "-map", "0",
            "-c", "copy",
            "-f", "segment",
            "-segment_time", str(SEGMENT_DURATION_SECS),
            "-segment_format", "mpegts",
            "-segment_start_number", str(start_num),
            pattern,
        ]
    args += [
        "-map", "0:v", "-c:v", "copy",
        "-f", "h264", "pipe:1",
        "-map", "0:a",
        "-c:a", "libopus", "-b:a", "32k",
        "-application", "lowdelay",
        "-frame_duration", "20",
        "-f", "ogg", "pipe:{audio_fd}",
    ]

    proc, audio_pipe = await _spawn_ffmpeg_with_audio_pipe(args, has_audio=True)
    logger.info("ffmpeg test file loop started (pid=%d)", proc.pid)

    tasks = [
        asyncio.create_task(_copy_to_live(proc.stdout, live_writer)),  # type: ignore[arg-type]
        asyncio.create_task(_drain_stderr(proc, "ffmpeg")),
    ]
    if audio_pipe is not None:
        tasks.append(asyncio.create_task(_audio_reader_task(audio_pipe, live_writer)))

    try:
        rc = await proc.wait()
    finally:
        for t in tasks:
            t.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)

    if rc != 0:
        raise RuntimeError(f"ffmpeg exited with code {rc}")


# --- real rpicam-vid + ffmpeg ---


async def _run_real_pipeline(
    cfg: CameraConfig,
    pattern: str,
    start_num: int,
    live_writer: LiveWriter,
) -> None:
    has_audio = not cfg.no_audio
    record = _records_segments(cfg)
    logger.info(
        "starting real capture (record=%s, has_audio=%s, segment_start=%d)",
        record, has_audio, start_num,
    )

    rpicam = await asyncio.create_subprocess_exec(
        "rpicam-vid",
        "--codec", "h264",
        "--inline",
        "--width", str(cfg.video_width),
        "--height", str(cfg.video_height),
        "--framerate", str(cfg.video_fps),
        "--bitrate", str(cfg.video_bitrate),
        "-t", "0",
        "-o", "-",
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        preexec_fn=os.setsid,
    )
    logger.info("rpicam-vid started (pid=%d)", rpicam.pid)

    rpicam_tasks = [asyncio.create_task(_drain_stderr(rpicam, "rpicam-vid"))]

    # Streaming-only with no audio: pipe rpicam straight to live writer.
    if not record and not has_audio:
        try:
            await _copy_to_live(rpicam.stdout, live_writer)  # type: ignore[arg-type]
            rc = await rpicam.wait()
        finally:
            for t in rpicam_tasks:
                t.cancel()
            await asyncio.gather(*rpicam_tasks, return_exceptions=True)
        if rc != 0:
            raise RuntimeError(f"rpicam-vid exited with code {rc}")
        return

    feed_video_to_ffmpeg = record
    alsa_idx = 1 if feed_video_to_ffmpeg else 0

    args = [
        "ffmpeg",
        "-nostdin", "-loglevel", "warning",
        "-probesize", "5M", "-analyzeduration", "5M",
    ]
    if feed_video_to_ffmpeg:
        args += ["-f", "h264", "-framerate", str(cfg.video_fps), "-i", "pipe:0"]
    if has_audio:
        device = cfg.audio_device or "default"
        args += ["-f", "alsa", "-i", device]

    if record:
        if has_audio:
            args += ["-map", "0:v", "-map", "1:a",
                     "-c:v", "copy", "-c:a", "aac", "-b:a", "64k"]
        else:
            args += ["-map", "0:v", "-c:v", "copy"]
        args += [
            "-f", "segment",
            "-segment_time", str(SEGMENT_DURATION_SECS),
            "-segment_format", "mpegts",
            "-segment_start_number", str(start_num),
            "-reset_timestamps", "1",
            pattern,
        ]

    if has_audio:
        args += [
            "-map", f"{alsa_idx}:a",
            "-c:a", "libopus", "-b:a", "32k",
            "-application", "lowdelay",
            "-frame_duration", "20",
            "-f", "ogg", "pipe:{audio_fd}",
        ]

    ffmpeg, audio_pipe = await _spawn_ffmpeg_with_audio_pipe(args, has_audio=has_audio)
    logger.info("ffmpeg started (pid=%d)", ffmpeg.pid)

    aux_tasks: list[asyncio.Task[None]] = [
        asyncio.create_task(_drain_stderr(ffmpeg, "ffmpeg")),
    ]
    if audio_pipe is not None:
        aux_tasks.append(asyncio.create_task(_audio_reader_task(audio_pipe, live_writer)))

    # Tee rpicam stdout: ffmpeg stdin + live writer (or just live writer
    # when not recording). Streaming-only mode + audio uses the live
    # writer alone for video; ffmpeg only sees ALSA.
    sinks: list[asyncio.StreamWriter | None] = []
    if feed_video_to_ffmpeg and ffmpeg.stdin is not None:
        sinks.append(ffmpeg.stdin)
    aux_tasks.append(asyncio.create_task(
        _copy_to_two(rpicam.stdout, sinks, live_writer)  # type: ignore[arg-type]
    ))

    try:
        rpicam_rc, ffmpeg_rc = await asyncio.gather(rpicam.wait(), ffmpeg.wait())
    finally:
        for t in [*rpicam_tasks, *aux_tasks]:
            t.cancel()
        await asyncio.gather(*rpicam_tasks, *aux_tasks, return_exceptions=True)

    if rpicam_rc != 0:
        raise RuntimeError(f"rpicam-vid exited with code {rpicam_rc}")
    if ffmpeg_rc != 0:
        raise RuntimeError(f"ffmpeg exited with code {ffmpeg_rc}")
